Dump basic block ids instead of pointers

mikedn opened the pull request:
Pointers are problematic in dump diffs...

mikedn commented:
@dotnet-bot test
Ubuntu16.04 arm Cross Debug Build
Tizen armel Cross Debug Build

mikedn commented:
@dotnet-bot test Ubuntu arm Cross Release Build

BruceForstall commented on src/jit/flowgraph.cpp:
```diff
@@ -7861,8 +7861,8 @@ void Compiler::fgAddSyncMethodEnterExit()
         {
-            printf("\nSynchronized method - Add 'acquired' initialization in first block BB%02u [%08p]\n", fgFirstBB,
-                   dspPtr(fgFirstBB));
+            printf("\nSynchronized method - Add 'acquired' initialization in first block BB%02u [%04u]\n", fgFirstBB,
```
Looks like an existing bug here where the first arg should be `fgFirstBB->bbNum`

mikedn replied:
Right, I'll fix

BruceForstall commented:
Are you familiar with `COMPlus_JitDiffableDasm=1`? Wouldn't that have the same effect? We already dump block numbers (or should in all cases), which are diffable. And if we need "diffable-ness", we use dspPtr that works with `COMPlus_JitDiffableDasm`. I guess the block id concept would work across block renumberings, and still be diffable.

mikedn commented:
> Are you familiar with COMPlus_JitDiffableDasm=1? 

Yes but the way that usually works is:
* "oh crap, I forgot to set it to true, I need to start over"
* "oh crap, I wish I hadn't set that, now everything is d1ffab1e and I can't search anything"

Note that I intentionally print the block id using only 4 digits while tree ids are printed using 6 digits. So if you search for [0001] you'll find blocks and if you search for [000001] you'll find trees.

>  I guess the block id concept would work across block renumberings, and still be diffable.

Yep. Here's a renumbering example from a rather trivial method dump:
```
BBnum         descAddr ref try hnd preds           weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1     [000..00E)-> BB04 ( cond )                     i label target gcsafe 
BB03 [0003]  1       BB01                  0.50  [???..???)                                     i 
BB04 [0001]  2       BB01,BB03             1     [00E..010)        (return)                     i label target gcsafe 
--------------------------------------------------------------------------------------------------------------------------------------

...

--------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1     [000..00E)-> BB03 ( cond )                     i label target gcsafe 
BB02 [0003]  1       BB01                  0.50  [???..???)                                     i 
BB03 [0001]  2       BB01,BB02             1     [00E..010)        (return)                     i label target gcsafe 
--------------------------------------------------------------------------------------------------------------------------------------
```

BruceForstall commented:
I like the idea.

I'm a little surprised we don't have methods on BasicBlock to create the block string, e.g.:
```
#ifdef DEBUG
   const char* dspToString()
   {
       static char buf[3][2000]; // static array of 3 to allow 3 concurrent calls in one printf()
       static int buf_index = 0;
       sprintf(buf[buf_index], 2000, "BB%02u [%04u]", bbNum, dspBlockID(this));
       return buf[buf_index];
   }
   void dspBlock()
   {
       printf("%s", dspToString());
   }
#endif // DEBUG
```

then if everywhere used block->dspToString(), you could add `dspPtr(this)` in one place if you wanted to see it everywhere.

mikedn commented:
> const char* dspToString()

That would be nice but the use of statics is fishy. How exactly does it allow 3 concurrent calls? Looks like buf_index is always 0.

Maybe we should just allocate like RangeCheck does in Range:
```C++
char* ToString(IAllocator* alloc)
{
    size_t size = 64;
    char*  buf  = (char*)alloc->Alloc(size);
    sprintf_s(buf, size, "<%s, %s>", lLimit.ToString(alloc), uLimit.ToString(alloc));
    return buf;
}
```
It uses more memory but we don't risk getting some garbled output.

BruceForstall commented:
Oops; forgot to increment, e.g.:
```
char* ret = buf[buf_index];
buf_index = (buf_index + 1) % 3;
return ret;
```
It's kind of a hack to avoid too much memory allocation, because we dump out these strings *A LOT*. I randomly chose 2000, but we know the size will be only about 20 (we could statically determine the max size).

And the static array is to allow:
```
printf("blk1 %s blk2 %s blk3 %s", blk1->dspToString(), blk2->dspToString(), blk3->dspToString());
```
where the three here are all live concurrently. "3" is a random number; could easily be 10 or 20 and still be better than a new allocation for every string dump.

mikedn commented:
Odd test failure:
```
D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\TestWrappers\JIT.Directed\JIT.Directed.XUnitWrapper.cs(24957): error : JIT_Directed._nullabletypes_isinstvaluetype_do_isinstvaluetype_do_._nullabletypes_isinstvaluetype_do_isinstvaluetype_do_cmd [FAIL] [D:\j\workspace\debug_windows---17180f1b\tests\runtest.proj]
        
  
  Return code:      1
  Raw output file:      D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\Reports\JIT.Directed\nullabletypes\isinstvaluetype_do\isinstvaluetype_do.output.txt
  Raw output:
  BEGIN EXECUTION
         "D:\j\workspace\debug_windows---17180f1b\tests\..\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root\corerun.exe" isinstvaluetype_do.exe 
        Object reference not set to an instance of an object.
        Test FAILED
        Expected: 100
        Actual: 666
        END EXECUTION - FAILED
        FAILED
        Test Harness Exitcode is : 1
        
  To run the test:
  > set CORE_ROOT=D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\Tests\Core_Root
  > D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\JIT\Directed\nullabletypes\isinstvaluetype_do\isinstvaluetype_do.cmd
  
        Expected: True
        Actual:   False
          D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\TestWrappers\JIT.Directed\JIT.Directed.XUnitWrapper.cs(24957,0): at JIT_Directed._nullabletypes_isinstvaluetype_do_isinstvaluetype_do_._nullabletypes_isinstvaluetype_do_isinstvaluetype_do_cmd()
```

BruceForstall commented:
I like the look of the change.

RE: the test failure -- It looks like all the CI tests passed. Was the test failure on your machine?

Is this still WIP? Are you planning to convert more BB print, or let that happen later/more incrementally?

mikedn commented:
> the test failure -- It looks like all the CI tests passed. Was the test failure on your machine?

That happened in a previous CI build. I rebased and pushed again and now it seems fine. It's odd because I don't remember that test failing before. There are a few tests that fail occasionally but I've never seen this one.

> Is this still WIP? Are you planning to convert more BB print, or let that happen later/more incrementally?

I think it's done, I've replaced existing code with `dspToString` in all relevant places that I have found. I didn't add `dspBlock` because I didn't find any suitable uses.

Anyway, I'm on holiday right now and I can't do much about this. Especially considering how "well" the Intenet works here :)

mikedn commented:
> I think it's done, I've replaced existing code with dspToString in all relevant places that I have found. I didn't add dspBlock because I didn't find any suitable uses.

To clarify things a bit: 

All the code that displayed block pointers has been changed to display block ids via `dspToString`. However, there's a lot of code that displays only the block number. That code is unchanged. 

Perhaps it would be useful to be consistent and always display the block id as well. Or, considering that block numbers change, maybe we should display only the block id most of the time. Opinions?

BruceForstall commented:
IMO, we probably want to continue displaying block numbers, since some code uses them to make decisions, e.g., for determining whether a branch is backwards or forwards.

mikedn commented:
OK, then this is good to go as is. 

Here's a before/after dump diff: https://gist.github.com/mikedn/35c481a121d41a00d2f9bbf775d80ee1

BruceForstall commented:
@mikedn Looks like you need to alter the block dump header so columns line up again:
```
 --------------------------------------------------------------------------------------------------------------------------------------
 BBnum         descAddr ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
 --------------------------------------------------------------------------------------------------------------------------------------
```

mikedn commented:
Oops, updated padding and column name:
```
--------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd                 weight    [IL range]     [jump]      [EH region]         [flags]
--------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1     [000..006)-> BB03 (always)                     
```

BruceForstall commented:
hmmmm... timeout
```
15:58:22   cmdLine:D:\j\workspace\debug_windows---17180f1b\bin\tests\Windows_NT.x64.Debug\JIT\Methodical\tailcall_v4\hijacking\hijacking.cmd Timed Out
```
try again?

@dotnet-bot test Windows_NT x64 Debug Build and Test
